XpressAI/xircuits
# ✨ Xircuits Context

# Description

This pull request introduces xircuits context (ctx),which allows users to pass data to another components in the form of a python dict. 
Previously to pass any type of data, the user needs to be explicitly code it via outPort to inPorts. This would be cumbersome especially if it's a global variable that might be used everywhere, such as the experiment name. By using ctx, users simply need to add it in their execute section. Here is a basic syntax:

```
@xai_component
class HelloContext(Component):

    def __init__(self, ctx):

    def execute(self, ctx) -> None:

        context_dict = {"new ctx": "Hello Xircuits!"}
        ctx.update(context_dict)
```

I have added the `HelloContext` component to the example components library for you to try.


## Pull Request Type


## Type of Change


# Tests



1. **Ensure that ctx can be accessed by all components**

    1. Drag 2 `Hello_Context" and link them together.
    2. The context should be accessible in the next component

You may also try out this [xircuit](https://gist.github.com/MFA-X-AI/eb5e77969e0c82c90c12276e526efbef).

Expected Output:

![image](https://user-images.githubusercontent.com/68586800/155682132-e252d1a3-1262-4330-9998-1308cd55f9ba.png)

2. **Ensure that previous components are still working**

    1. Run any Xircuits examples. Shouldn't get any errors.
    2. Check the github E2E github action. Should pass the test.


## Tested on?


# Notes

As the ctx parameter needs to be passed in, I had to update every component in this pull request. You will need to pull a fresh copy / not use your old component library.
